### PR TITLE
Refactor tests

### PR DESCRIPTION
**Pull request by nwu63**
## Purpose
This PR refactors the existing tests by introducing a test class that handles various shared tasks
- run an optimization
- check that solutions are close
- check that the final inform is correct
- check the history file contains expected contents
- check hot start

I also added parameterized tests etc. for efficiency.

Not all existing tests use this class however:
- the two SNOPT test files can probably use this if we want, but will require a bit more work
- for some reason NSGA2 will segfault if the `objfunc` is defined within the class. Absolutely no idea why..

I'll keep this as a draft PR for now. The main thing I want to check is that we're not losing any coverage through this, so if reviewers can also check against the old tests and see if I missed anything. Finally, the way the tests are structured can probably improved, so any suggestions are welcome.

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [x] Maintenance update
- [ ] Other (please describe)

### CODE REVIEW

**Review by sseraj — COMMENTED** [review]
This is a major improvement! I don't see any obvious reduction in coverage other than SNOPT versions prior to 7.7.7. Is that intentional?

File: test/test_large_snopt.py [hunk 1]
```diff
@@ -59,83 +67,68 @@ def sens(self, xdict, funcs):
...
-            "Print file": "{}.out".format(test_name),
-            "Summary file": "{}_summary.out".format(test_name),
-        }
-        self.optimize(tol=1e-5, optOptions=optOptions)
+        self.optName = "SNOPT"
+        self.setup_optProb(sparse=True)
+        sol = self.optimize()
+        self.assert_solution_allclose(sol, 1e-5, partial_x=True)
+
+    def test_sparse_IPOPT(self):
```
[the remark this answers is not in the document]
Comment (marcomangano) [reply test/test_large_snopt.py:112]: yeah I agree, can't we use @parameterized(["SNOPT", "IPOPT"]) for `test_sparse` @nwu63 ?

File: test/test_hs015.py [hunk 2]
```diff
@@ -53,190 +78,32 @@ def sens(self, xdict, funcs):
...
-
-        self.xStar1 = (0.5, 2.0)
-        self.xStar2 = (-0.79212322, -1.26242985)
-
-        dv = sol.getDVs()
-        sol_xvars = [sol.variables["xvars"][i].value for i in range(2)]
-        assert_allclose(sol_xvars, dv["xvars"], atol=tol, rtol=tol)
-        # we check either optimum via try/except
-        try:
-            if optName == "SNOPT" and opt.version != "7.7.7":
```
[the remark this answers is not in the document]
Comment (nwu63) [reply test/test_hs015.py:110]: Good point, will add

File: test/test_hs015.py [hunk 3]
```diff
@@ -249,34 +116,20 @@ def test_snopt(self):
...
-            "IFILE": "hs015_CONMIN.out",
-        }
-        self.optimize_with_hotstart("CONMIN", 1e-10, optOptions=optOptions)
-
-    def test_psqp(self):
-        optOptions = {"IFILE": "hs015_PSQP.out"}
-        self.optimize_with_hotstart("PSQP", 5e-12, optOptions=optOptions)
+        # dv = sol.getDVs()
+        # sol_xvars = [sol.variables["xvars"][i].value for i in range(2)]
+        # assert_allclose(sol_xvars, dv["xvars"], atol=tol, rtol=tol)
```
[the remark this answers is not in the document]
Comment (nwu63) [reply test/test_hs015.py:121]: Ah, right these were testing `sol.variables` etc. I will try to generalize these and make sure the Solution object contained the right values in different attributes.

File: test/test_large_snopt.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nwu63) [reply test/test_large_snopt.py:112]: Good point, will rename the file and parameterize those. Also Marco because you didn't wrap that in quotes you just notified some random person on GitHub :stuck_out_tongue: 

I added IPOPT because I don't think we have been testing its sparse Jacobian capabilities, but maybe the problem size is a bit unreasonable. Should we reduce the size?

[the remark this answers is not in the document]
Comment (sseraj) [reply test/test_large_snopt.py:112]: I think the size could be reduced. I just checked the Azure logs and this one test is about 50% of the 'Run Tests' time.

[the remark this answers is not in the document]
Comment (nwu63) [reply test/test_large_snopt.py]: For some reason, setting it to `5000` resulted in some segfaults within IPOPT, while 10000 is okay. Could you double check and see if you could reproduce that on your local machine?

[the remark this answers is not in the document]
Comment (sseraj) [reply test/test_large_snopt.py]: Yes, I was able to reproduce the segfault. That is very strange.

File: test/test_hs015.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nwu63) [reply test/test_hs015.py:121]: I added some tests, but not all since it's a bit harder to generalize. I've opened #256 to address this, among other things.

File: test/testing_utils.py [hunk 4]
```diff
@@ -25,12 +28,35 @@ def assert_dict_allclose(actual, desired, atol=DEFAULT_TOL, rtol=DEFAULT_TOL, pa
...
+    ------
+    AssertionError
+        if the values for any key do not match to the target tolerance.
     """
     if not partial:
-        assertEqual(set(actual.keys()), set(desired.keys()))
-    for key in actual.keys():
-        if key in desired.keys() or not partial:
-            assert_allclose(actual[key], desired[key], atol=atol, rtol=rtol, err_msg=f"Failed for key {key}")
+        assert set(actual.keys()) == set(desired.keys())
```
Comment (marcomangano) [inline test/testing_utils.py:53]: why did you drop `assertEqual` for this check?

**Review by marcomangano — COMMENTED** [review]
Altright, I think I listed all the remaining comments I had. Note that some of them are not necessarily related to your edits, but additional comments/info we could add during this refactoring.

This is a substantial PR and some commits are pretty big and hard to follow, so for future PRs I would suggest to either break them further or, more in general, add a comment to the commit message to clarify the rationale behind it?

File: test/testing_utils.py [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+            for x in self.xStar:
+                dist.append(get_dict_distance(x, sol.xStar))
+            self.sol_index = dist.index(min(dist))
+        else:
+            # assume we have a single solution
+            self.sol_index = 0
+        # now we assert against the closest solution
+        # objective
+        # sol.fStar was broken for earlier versions of SNOPT
+        if self.optName == "SNOPT" and Version(self.optVersion) >= Version("7.7.7"):
```
Comment (sseraj) [inline test/testing_utils.py:167]: What does `Version` do here? It introduces an additional dependency on `packaging` and I'm not sure it's necessary.

**Review by sseraj — COMMENTED** [review]
This looks good to me other than one thing

File: test/testing_utils.py (hunk 5 above)
Comment (nwu63) [reply test/testing_utils.py:167]: It allows for "correct" comparison of version strings, which is a non-trivial task (even though it's relatively simple here). `packaging` is a dependency of `setuptools` I believe, but I've switched to using `pkg_resources` which is something built into `setuptools` and therefore should available for everyone.

File: test/testing_utils.py (hunk 4 above)
Comment (nwu63) [reply test/testing_utils.py:53]: `assertEqual` cannot be used because this function is not inside a `unittest.TestCase` class.

File: test/testing_utils.py [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+def assert_optProb_size(optProb, nObj, nDV, nCon):
+    """Checks that nObj, nDV and nCon are correct for optProb"""
+    optProb.finalize()
+    assertEqual(optProb.nObj, nObj)
+    assertEqual(optProb.nCon, nCon)
+    assertEqual(optProb.ndvs, nDV)
+
+
+def get_dict_distance(d, d2):
+    """This converts a flat 1D dict to array using sorted keys, then computes the L2 distance"""
```
[the remark this answers is not in the document]
Comment (nwu63) [reply test/testing_utils.py:89]: Not sure what you mean here. The code is not 100% robust because the numpy array casts aren't 100% fool proof, but the approach itself should be valid no?

File: test/test_tp109.py [hunk 7]
```diff
@@ -164,59 +167,28 @@ def test_snopt(self):
...
-        sol2 = opt2(sol1)
+        # Now re-optimize with SNOPT
+        self.optProb = sol1
+        self.optName = "SNOPT"
+        self.setup_optProb()
+        sol2 = self.optimize(sens="CS")
 
         # Check Solution
-        assert_allclose(sol2.objectives["obj"].value, 0.536206927538e04, atol=1e-2, rtol=1e-2)
+        self.assert_solution(sol2, 1e-2)
```
[the remark this answers is not in the document]
Comment (nwu63) [reply test/test_tp109.py]: Hmm good point, I think the idea is that this is loose because SNOPT may not be available, so then we just assert the ALPSO answer. I will fix this.

File: test/test_tp109.py [hunk 8]
```diff
@@ -164,59 +167,28 @@ def test_snopt(self):
...
-
-        # Now run the previous solution with SNOPT
-        try:
-            opt2 = OPT("SNOPT")
-        except Error:
-            raise unittest.SkipTest("Optimizer not available:", "SNOPT")
+        sol1 = self.optimize(sens="CS")
 
-        sol2 = opt2(sol1)
+        # Now re-optimize with SNOPT
```
[the remark this answers is not in the document]
Comment (nwu63) [reply test/test_tp109.py:189]: Yes, we run ALPSO to get near the optimum, then use SNOPT to "refine" by cold starting from the best ALPSO point. This PR does not change any of the tests, this is something that has been there for ages, and it's good because it tests that the `Solution` object can be used to directly cold start optimizations. Although we should check that the DVs get transferred I guess..

File: test/test_hs071.py [hunk 9]
```diff
@@ -37,81 +66,30 @@ def sens(self, xdict, funcs):
...
     def test_slsqp_setDV(self):
         """
         Test that setDV works as expected, even with scaling/offset
         """
-        test_name = "hs071_SLSQP_setDV"
-        histFileName = "{}.hst".format(test_name)
+        self.optName = "SLSQP"
+        histFileName = "hs071_SLSQP_setDV.hst"
         newDV = {"xvars": np.array([1, 4, 4, 1])}
-        optOptions = {"IFILE": "{}.out".format(test_name)}
```
[the remark this answers is not in the document]
Comment (nwu63) [reply test/test_hs071.py:103]: The output file names are uniformly handled in the `OptTest` class, which are generated based on the name of the individual tests. See [here](https://github.com/mdolab/pyoptsparse/blob/update-tests/test/testing_utils.py#L114-L125) for the mapping of option name and suffixes, and [here](https://github.com/mdolab/pyoptsparse/blob/update-tests/test/testing_utils.py#L203-L221) for how the options are set.

File: test/testing_utils.py [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+            The optimizer options dictionary
+
+        Returns
+        -------
+        optOptions: dict
+            The same optOptions dictionary, but with the output file names set.
+        """
+        optionName = OUTPUT_FILENAMES[self.optName]
+        for optionName, suffix in OUTPUT_FILENAMES[self.optName].items():
+            optOptions[optionName] = f"{self.name}_{self._testMethodName}{suffix}"
```
[the remark this answers is not in the document]
Comment (nwu63) [reply test/testing_utils.py:220]: Yeah I know what you mean, these are coming from `unittest`. Will do.

File: test/test_hs071.py [hunk 11]
```diff
@@ -75,12 +74,11 @@ def test_slsqp_setDV(self):
         """
         Test that setDV works as expected, even with scaling/offset
         """
-        test_name = "hs071_SLSQP_setDV"
-        histFileName = "{}.hst".format(test_name)
+        self.optName = "SLSQP"
```
[the remark this answers is not in the document]
Comment (nwu63) [reply test/test_hs071.py:87]: I did it because I didn't want to pass the same variable around everywhere (it's needed in a lot of places). I think it's okay but would like to hear some feedback on this. @sseraj any thoughts?

Yeah `self.opt` created some obscure pickling error, took me a while to figure it out (still not really sure why that was the case). In any case storing the optimizer version was easy enough.

After looking at the code, I will add some more documentation and also set placeholder attributes in the class, otherwise it's unclear what attributes are needed to use `OptTest`.

File: test/testing_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (nwu63) [reply test/testing_utils.py]: I ended up changing these to `self.id()` since that's a public API so hopefully more stable. It's also more explicit even though it will generate some pretty long file names.

File: test/test_tp109.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nwu63) [reply test/test_tp109.py]: Actually ALPSO seems...kinda broken. I'll open an issue but I prefer not fixing this in this PR.

**Comment by nwu63** [comment]
> This is a substantial PR and some commits are pretty big and hard to follow, so for future PRs I would suggest to either break them further or, more in general, add a comment to the commit message to clarify the rationale behind it?

Yeah sorry about that, I tried to break it up but as you can see, it's hard to do when I'm changing many things at once. Because I will be testing things as I develop, it's hard to keep a nice linear git history as well. I will try to make my commit messages better though, I don't think I was very good at that.

Anyways, I think I addressed most of the concerns.

File: test/test_hs071.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (marcomangano) [reply test/test_hs071.py:103]: Oh gotcha, I saw that but removed the info right away

File: test/testing_utils.py (hunk 4 above)
Comment (marcomangano) [reply test/testing_utils.py:53]: Makes sense, did not see that

File: test/testing_utils.py (hunk 5 above)
Comment (sseraj) [reply test/testing_utils.py]: Thanks

File: test/test_hs071.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (sseraj) [reply test/test_hs071.py:87]: I can see it being clearer to pass as an argument, but I think it's fine as is because you added the docstring and this is just used for testing.

**Review by sseraj — APPROVED** [review]

File: test/testing_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcomangano) [reply test/testing_utils.py:89]: Yeah I was just thinking about cases where the L2 norm would not be the the best approach but I think for testing specifically this is not an issue. Good to go!

**Review by marcomangano — APPROVED** [review]
